src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs

```diff
@@ -180,6 +208,156 @@ public void Pop(bool success)
             }
         }
 
+        // asynchronously await any pending stacks that resumable converters depend on.
+        public async ValueTask AwaitPendingTask()
+        {
+            Debug.Assert(PendingTask != null);
+
+            if (!PendingTask.IsCompleted)
+            {
+                // wrap with Task.WhenAny to avoid surfacing any exceptions here
+                await Task.WhenAny(PendingTask).ConfigureAwait(false);
+            }
+
+            // Do not clear the `PendingTask` field here since the result
+            // will need to be consumed by a resumable converter.
+        }
+
+        // Asynchronously dispose of any AsyncDisposables that have been scheduled for disposal
+        public async ValueTask DisposePendingAsyncDisposables()
+        {
+            Debug.Assert(PendingAsyncDisposables?.Count > 0);
+            List<Exception>? exceptions = null;
+
+            foreach (IAsyncDisposable asyncDisposable in PendingAsyncDisposables)
+            {
+                try
+                {
+                    await asyncDisposable.DisposeAsync().ConfigureAwait(false);
+                }
+                catch (Exception exn)
+                {
+                    exceptions ??= new();
+                    exceptions.Add(exn);
+                }
+            }
+
+            if (exceptions is not null)
+            {
+                if (exceptions.Count == 1)
+                {
+                    ExceptionDispatchInfo.Capture(exceptions[0]).Throw();
+                }
+
+                throw new AggregateException(exceptions);
```

Comment:
This is interesting.  FWIW, we faced a similar dilemma back in .NET 4.5 with await and ultimately decided to just throw one even if there's multiple.  One of the primary drivers was concern around non-determinism, and whether folks might write catch blocks for just the target exception and not realize they need to handle an aggregate as well, and then even if they did realize it, end up needing to duplicate all of their exception handling logic to handle both the single and aggregate cases.

Reply:
> One of the primary drivers was concern around non-determinism, and whether folks might write catch blocks for just the target exception and not realize they need to handle an aggregate as well, and then even if they did realize it, end up needing to duplicate all of their exception handling logic to handle both the single and aggregate cases.

I can definitely see where that's coming from, but I wonder if suppressing exception information is desireable in principle: it could be argued, especially if you subscribe to the "let it crash" school of thought, that logging exceptions is more important than handling them.

In this particular case it probably doesn't matter, I don't think STJ users would or should try to handle this particular exception. It's meant to address a corner case of a corner case, where there is likely something wrong in their `I(Async)Disposable` implementation, so reporting back with as much information as possible seems appropriate.

Reply:
Ok, changing it to only return one exception.